File: frontend/src/lib/pos.ts
const BASE = "/api/v1/pos";

export type POSSessionStatus = "OPEN" | "CLOSED";
export type POSPaymentMethod = "CASH" | "MPESA" | "CARD" | "SPLIT";
export type POSSaleStatus    = "COMPLETED" | "VOIDED";

export interface POSSaleLineCreate {
  product_id:   string;
  qty:          number;
  unit_price:   number;
  discount_pct: number;
  tax_rate:     number;
}

export interface POSSaleLine extends POSSaleLineCreate {
  id:           string;
  line_no:      number;
  product_name: string | null;
  product_sku:  string | null;
  line_total:   number;
}

export interface POSSale {
  id:             string;
  sale_no:        string;
  session_id:     string;
  cashier_id:     string | null;
  cashier_name:   string | null;
  customer_id:    string | null;
  customer_name:  string | null;
  payment_method: POSPaymentMethod;
  status:         POSSaleStatus;
  subtotal:       number;
  discount_total: number;
  tax_total:      number;
  sale_total:     number;
  amount_paid:    number | null;
  change_given:   number | null;
  mpesa_ref:      string | null;
  notes:          string | null;
  created_at:     string;
  lines:          POSSaleLine[];
}

export interface POSSession {
  id:                 string;
  session_no:         string;
  cashier_id:         string | null;
  cashier_name:       string | null;
  register_id:        string;
  status:             POSSessionStatus;
  opened_at:          string;
  closed_at:          string | null;
  opening_float:      number;
  closing_cash:       number | null;
  expected_cash:      number | null;
  cash_difference:    number | null;
  total_sales:        number;
  total_transactions: number;
  notes:              string | null;
  created_at:         string;
}

export interface POSDashboard {
  today_sales_count:  number;
  today_sales_total:  number;
  today_cash_total:   number;
  today_mpesa_total:  number;
  today_card_total:   number;
  open_sessions:      number;
  avg_basket_value:   number | null;
}

async function req<T>(url: string, opts?: RequestInit): Promise<T> {
  const r = await fetch(url, { credentials: "include", ...opts });
  if (!r.ok) {
    const err = await r.json().catch(() => ({ detail: r.statusText }));
    throw new Error(err.detail ?? r.statusText);
  }
  return r.json();
}

const json = (body: unknown, method = "POST") => ({
  method,
  headers: { "Content-Type": "application/json" },
  body: JSON.stringify(body),
});

export const posApi = {
  dashboard: () => req<POSDashboard>(`${BASE}/dashboard`),

  currentSession: () => req<POSSession | null>(`${BASE}/sessions/current`),
  listSessions: (status?: POSSessionStatus) => {
    const q = status ? `?status=${status}` : "";
    return req<POSSession[]>(`${BASE}/sessions${q}`);
  },
  openSession: (register_id: string, opening_float: number) =>
    req<POSSession>(BASE + "/sessions", json({ register_id, opening_float })),
  closeSession: (id: string, closing_cash: number, notes?: string) =>
    req<POSSession>(`${BASE}/sessions/${id}/close`, json({ closing_cash, notes })),

  createSale: (body: {
    customer_id?: string;
    payment_method: POSPaymentMethod;
    amount_paid?: number;
    mpesa_ref?: string;
    lines: POSSaleLineCreate[];
  }) => req<POSSale>(BASE + "/sales", json(body)),

  listSales: () => req<POSSale[]>(`${BASE}/sales`),
  voidSale: (id: string) => req<POSSale>(`${BASE}/sales/${id}/void`, { method: "POST" }),
};

export function fmtKES(v: number) {
  return new Intl.NumberFormat("en-KE", { style: "currency", currency: "KES", minimumFractionDigits: 2 }).format(v);
}

File: backend/app/schemas/pos.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Optional, List
import uuid

from pydantic import BaseModel, ConfigDict

from app.models.pos import POSSessionStatus, POSPaymentMethod, POSSaleStatus


class POSSaleLineCreate(BaseModel):
    product_id:  uuid.UUID
    qty:         Decimal
    unit_price:  Decimal
    discount_pct: Decimal = Decimal("0")
    tax_rate:    Decimal = Decimal("16")


class POSSaleLineRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:           uuid.UUID
    line_no:      int
    product_id:   uuid.UUID
    product_name: Optional[str] = None
    product_sku:  Optional[str] = None
    qty:          Decimal
    unit_price:   Decimal
    discount_pct: Decimal
    tax_rate:     Decimal
    line_total:   Decimal


class POSSaleCreate(BaseModel):
    customer_id:    Optional[uuid.UUID]   = None
    payment_method: POSPaymentMethod      = POSPaymentMethod.CASH
    amount_paid:    Optional[Decimal]     = None
    mpesa_ref:      Optional[str]         = None
    notes:          Optional[str]         = None
    lines:          List[POSSaleLineCreate]


class POSSaleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:             uuid.UUID
    sale_no:        str
    session_id:     uuid.UUID
    cashier_id:     Optional[uuid.UUID]
    cashier_name:   Optional[str] = None
    customer_id:    Optional[uuid.UUID]
    customer_name:  Optional[str] = None
    payment_method: POSPaymentMethod
    status:         POSSaleStatus
    subtotal:       Decimal
    discount_total: Decimal
    tax_total:      Decimal
    sale_total:     Decimal
    amount_paid:    Optional[Decimal]
    change_given:   Optional[Decimal]
    mpesa_ref:      Optional[str]
    notes:          Optional[str]
    created_at:     datetime
    lines:          List[POSSaleLineRead] = []


class OpenSessionRequest(BaseModel):
    register_id:   str = "REGISTER-1"
    opening_float: Decimal = Decimal("0")


class CloseSessionRequest(BaseModel):
    closing_cash: Decimal
    notes:        Optional[str] = None


class POSSessionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id:                uuid.UUID
    session_no:        str
    cashier_id:        Optional[uuid.UUID]
    cashier_name:      Optional[str] = None
    register_id:       str
    status:            POSSessionStatus
    opened_at:         datetime
    closed_at:         Optional[datetime]
    opening_float:     Decimal
    closing_cash:      Optional[Decimal]
    expected_cash:     Optional[Decimal]
    cash_difference:   Optional[Decimal]
    total_sales:       Decimal
    total_transactions: int
    notes:             Optional[str]
    created_at:        datetime


class POSDashboard(BaseModel):
    today_sales_count:  int
    today_sales_total:  Decimal
    today_cash_total:   Decimal
    today_mpesa_total:  Decimal
    today_card_total:   Decimal
    open_sessions:      int
    avg_basket_value:   Optional[Decimal]
